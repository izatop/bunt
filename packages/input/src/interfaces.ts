import {MaybeArray, MayNullable} from "@bunt/util";
import {Fields, List} from "./Type";
import {TypeAbstract} from "./TypeAbstract";

export type Scalars = string | number | boolean;

export interface Compound {
    [key: string]: MaybeArray<Scalars | Compound>;
}

export type MayInput = MayNullable<MaybeArray<MayNullable<Scalars | Compound>>>;
export type MayListInput = MayNullable<Scalars | Compound>;

export type FieldFn<T> = () => T;
export type FieldType<T> = T | FieldFn<T>;

export type FieldsSchema<T> = {
    [K in keyof T]-?: T[K] extends Array<infer S>
        ? FieldType<List<S, any>>
        : T[K] extends Record<any, any>
            ? FieldType<Fields<T[K]>>
            : FieldType<TypeAbstract<T[K], any>>;
};

export type ObjectFields<T> = T extends Promise<infer A>
    ? FieldsSchema<Exclude<A, undefined | null>>
    : FieldsSchema<Exclude<T, undefined | null>>;

export type FieldSelectType<T> = T extends Record<string, any>
    ? FieldType<Fields<T>>
    : T extends Array<infer I>
        ? FieldType<List<I, any>>
        : FieldType<TypeAbstract<T, any>>;

export type ObjectTypeMerge<T extends Record<string, any>> = Fields<T> | ObjectFields<T>;
