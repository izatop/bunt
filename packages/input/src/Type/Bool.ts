import {isBoolean} from "@typesafeunit/util";
import {ScalarType} from "./ScalarType";

export const Bool = new ScalarType<boolean, boolean>({
    name: "Bool",
    validate(payload) {
        this.assert(isBoolean(payload), `Wrong payload: ${this.name} expected`, payload);
        return payload;
    },
});
