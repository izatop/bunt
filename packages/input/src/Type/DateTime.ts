import {isNumber, isString} from "@bunt/util";
import {ScalarType} from "./ScalarType";

export const DateTime = new ScalarType<Date, string | number>({
    name: "DateTime",
    validate(payload) {
        this.assert(isNumber(payload) || isString(payload), `Wrong payload: ${this.name} expected`, payload);
        this.assert(validateDateValue(payload), "Wrong date format", payload);

        return new Date(payload);
    },
});

function validateDateValue(payload: string | number): boolean {
    try {
        new Date(payload).toISOString();
    } catch {
        return false;
    }

    return true;
}
