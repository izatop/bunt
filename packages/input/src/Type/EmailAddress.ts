import {isString} from "@bunt/util";
import {ScalarType} from "./ScalarType";

// eslint-disable-next-line
export const RE_EMAIL_RFC5222 = /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/;

export const EmailAddress = new ScalarType<string>({
    name: "EmailAddress",
    validate(payload): string {
        this.assert(isString(payload), `Wrong payload: ${this.name} expected`, payload);
        this.assert(RE_EMAIL_RFC5222.test(payload), `Wrong payload: ${this.name} expected`, payload);

        return payload;
    },
});
