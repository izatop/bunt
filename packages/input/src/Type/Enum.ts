import {isString, Promisify} from "@bunt/util";
import {TypeAbstract} from "../TypeAbstract.js";

export class Enum<T extends string | number> extends TypeAbstract<T> {
    readonly #value: any;

    constructor(value: {[key: string]: T}) {
        super();
        this.#value = value;
    }

    public validate(input: unknown): Promisify<T> {
        this.assert(isString(input), "Wrong type", input);
        this.assert(!/^\d+$/.test(input), "Wrong value", input);
        this.assert(input in this.#value, "Wrong value", input);

        return this.#value[input];
    }
}
