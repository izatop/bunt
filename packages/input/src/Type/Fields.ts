import {entriesReverse, isFunction, isInstanceOf, isObject, toError} from "@bunt/util";
import {AssertionObjectError, AssertionTypeError, IReadableTypeError} from "../Assertion";
import {ObjectFields, ObjectTypeMerge} from "../interfaces";
import {TypeAbstract} from "../TypeAbstract";

export class Fields<TValue extends Record<string, any>> extends TypeAbstract<TValue> {
    readonly #fields: ObjectFields<TValue>;
    readonly #name: string;

    constructor(fields: ObjectFields<TValue>, name = "Object") {
        super();
        this.#fields = fields;
        this.#name = name;
    }

    public get name(): string {
        return this.#name;
    }

    public get fields(): ObjectFields<TValue> {
        return this.#fields;
    }

    public merge<F extends Record<string, any>>(from: ObjectTypeMerge<F>): Fields<TValue & F> {
        if (isInstanceOf(from, Fields)) {
            return new Fields<TValue & F>(
                Object.assign({}, this.fields, from.fields) as any,
            );
        }

        return new Fields<TValue & F>(
            Object.assign({}, this.fields, from) as any,
        );
    }

    public async validate(payload: unknown): Promise<TValue> {
        this.assert(isObject(payload), `Wrong payload: ${this.name} expected`, payload);

        const entries: [string, any][] = [];
        const validations = new Map<string, IReadableTypeError>();
        for (const [field, type] of Object.entries(this.#fields)) {
            try {
                const validationType = isFunction(type) ? type() : type;
                this.assert(
                    isInstanceOf(validationType, TypeAbstract),
                    `Wrong field: ${field}`,
                    payload[field],
                );

                entries.push([field, await validationType.validate(payload[field])]);
            } catch (error) {
                if (isInstanceOf(error, AssertionTypeError)) {
                    validations.set(field, error.toSafeJSON());
                    continue;
                }

                validations.set(field, {
                    payload: payload[field],
                    message: toError(error, "Unknown").message,
                    type: "Unknown",
                });
            }
        }

        if (validations.size > 0) {
            throw new AssertionObjectError(
                "Assertion failed",
                this,
                payload,
                entriesReverse([...validations.entries()]),
            );
        }

        return entriesReverse(entries);
    }
}
