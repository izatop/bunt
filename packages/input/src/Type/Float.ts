import {isNumber} from "@bunt/util";
import {ScalarType} from "./ScalarType";

export const Float = new ScalarType<number>({
    name: "Float",
    validate(payload) {
        this.assert(isNumber(payload), `Wrong payload: ${this.name} expected`, payload);
        this.assert(
            payload <= Number.MAX_SAFE_INTEGER && payload >= Number.MIN_SAFE_INTEGER,
            "Wrong payload range", payload,
        );

        return payload;
    },
});
