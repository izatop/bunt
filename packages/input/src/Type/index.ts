export * from "./DateTime";
export * from "./Float";
export * from "./Int";
export * from "./ToNumber";
export * from "./List";
export * from "./NonNull";
export * from "./Nullable";
export * from "./NonNull";
export * from "./Fields";
export * from "./ObjectType";
export * from "./Text";
export * from "./Varchar";
export * from "./Union";
export * from "./Bool";
export * from "./UUID";
export * from "./JSONString";
