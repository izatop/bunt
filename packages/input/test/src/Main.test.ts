import {isInstanceOf} from "@bunt/util";
import {
    AssertionObjectError,
    Bool,
    DateTime,
    Fields,
    Float,
    Int,
    JSONString,
    List,
    NonNull,
    Nullable,
    Text,
    ToNumber,
    TypeAbstract,
    Union,
    validate,
    Varchar,
} from "../../src";
import {ITestDescription, ITestHobby, ITestType} from "./interfaces";
import {TestEnum, TestEnumType} from "./Type/TestEnum";

describe("Test Input", () => {
    const rand = Math.random();
    const union = new Union<Date | boolean>(
        (input) => {
            switch (typeof input) {
                case "string":
                case "number":
                    return DateTime;
                case "boolean":
                    return Bool;
            }
        },
    );

    const samples: [any, any, TypeAbstract<any>][] = [
        [1, 1, Int],
        ["1", 1, ToNumber],
        [false, false, Bool],
        [true, true, Bool],
        [rand, rand, Float],
        [null, undefined, new Nullable(Int)],
        [undefined, undefined, new Nullable(Int)],
        [null, "foo", new NonNull(Text, "foo")],
        [undefined, [1], new NonNull(new List(Int), [1])],
        ["text", "text", Text],
        ["text", "text", new Varchar({min: 0, max: 4})],
        [{v: 1, b: true, n: []}, {v: 1, b: true}, new Fields({v: Int, b: Bool})],
        [[1, 2, 3], [1, 2, 3], new List(Int)],
        [false, false, union],
        ["2020-01-01", new Date("2020-01-01"), union],
        [new Date("2020-01-01").getTime(), new Date("2020-01-01"), union],
        ["{\"foo\": \"bar\"}", {foo: "bar"}, JSONString],
        ["true", true, JSONString],
        ["1", 1, JSONString],
        ["[1]", [1], JSONString],
        ["STR", TestEnum.STR, TestEnumType],
    ];

    test.each(samples)(
        "Test %s -> %s",
        async (payload, expected, type) => {
            await expect(validate(type, payload)).resolves.toEqual(expected);
        },
    );

    test("AssertionError", async () => {
        const Description = new Fields<ITestDescription>({
            author: Text,
            date: DateTime,
            text: Text,
        });

        const Hobby = new Fields<ITestHobby>({
            type: Text,
            description: new Nullable(Description),
        });

        const human: Fields<ITestType> = new Fields<ITestType>({
            age: Int,
            name: Text,
            children: () => new NonNull(new List(human), []),
            parent: () => new Nullable(human),
            links: new List(Text),
            hobby: new Nullable(Hobby),
        }, "Human");

        const payload = {
            age: 32,
            parent: {name: "Parent", links: ["a"]},
            children: [
                {name: "Lisa", age: 8, links: ["b"]},
                {age: 12, name: "Bob", children: 1},
                {age: 3},
            ],
            links: ["a", "c"],
        };

        const pending = validate(human, payload);
        await expect(pending).rejects.toThrow();
        const error = await pending.catch((error) => error);
        expect(error).toBeInstanceOf(AssertionObjectError);
        if (isInstanceOf(error, AssertionObjectError)) {
            expect(error.getLogValue()).toMatchSnapshot();
        }
    });

    test("JSON parse fails", () => {
        expect(validate(JSONString, undefined))
            .rejects
            .toThrow();

        expect(validate(JSONString, "{123r1]"))
            .rejects
            .toThrow();
    });

    test("Test Enum", () => {
        expect(() => TestEnumType.validate("1")).toThrow();
        expect(() => TestEnumType.validate("str")).toThrow();
        expect(TestEnumType.validate("NUM")).toBe(TestEnum.NUM);
        expect(TestEnumType.validate("STR")).toBe(TestEnum.STR);
    });

    test("Merge", async () => {
        const foo = new Fields<{foo: number}>({foo: Int});
        const foobar = foo.merge(new Fields<{bar: number}>({bar: Int}));
        await expect(validate(foobar, {foo: 1, bar: 2}))
            .resolves
            .toEqual({foo: 1, bar: 2});
    });
});
