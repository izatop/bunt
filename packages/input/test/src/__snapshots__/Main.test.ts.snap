// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Input AssertionError 1`] = `
Object {
  "fields": Object {
    "children": Object {
      "fields": Array [
        Object {
          "fields": Object {
            "children": Object {
              "message": "Wrong payload: Human[] expected",
              "payload": 1,
              "type": "List",
            },
            "links": Object {
              "message": "Wrong payload: Text[] expected",
              "payload": undefined,
              "type": "List",
            },
          },
          "index": 1,
          "message": "Assertion failed",
          "payload": Object {
            "age": 12,
            "children": 1,
            "name": "Bob",
          },
          "type": "Human",
        },
        Object {
          "fields": Object {
            "links": Object {
              "message": "Wrong payload: Text[] expected",
              "payload": undefined,
              "type": "List",
            },
            "name": Object {
              "message": "Wrong payload: Text expected",
              "payload": undefined,
              "type": "Text",
            },
          },
          "index": 2,
          "message": "Assertion failed",
          "payload": Object {
            "age": 3,
          },
          "type": "Human",
        },
      ],
      "message": "Assertion failed",
      "payload": Array [
        Object {
          "age": 8,
          "links": Array [
            "b",
          ],
          "name": "Lisa",
        },
        Object {
          "age": 12,
          "children": 1,
          "name": "Bob",
        },
        Object {
          "age": 3,
        },
      ],
      "type": "List",
    },
    "name": Object {
      "message": "Wrong payload: Text expected",
      "payload": undefined,
      "type": "Text",
    },
    "parent": Object {
      "fields": Object {
        "age": Object {
          "message": "Wrong payload: Int expected",
          "payload": undefined,
          "type": "Int",
        },
      },
      "message": "Assertion failed",
      "payload": Object {
        "links": Array [
          "a",
        ],
        "name": "Parent",
      },
      "type": "Human",
    },
  },
  "message": "Assertion failed",
  "payload": Object {
    "age": 32,
    "children": Array [
      Object {
        "age": 8,
        "links": Array [
          "b",
        ],
        "name": "Lisa",
      },
      Object {
        "age": 12,
        "children": 1,
        "name": "Bob",
      },
      Object {
        "age": 3,
      },
    ],
    "links": Array [
      "a",
      "c",
    ],
    "parent": Object {
      "links": Array [
        "a",
      ],
      "name": "Parent",
    },
  },
  "type": "Human",
}
`;
